src/NHibernate/Type/TypeFactory.cs

```diff
@@ -522,6 +522,31 @@ public static IType HeuristicType(string typeName)
 			return HeuristicType(typeName, null);
 		}
 
+		/// <summary>
+		/// Uses heuristics to deduce a NHibernate type given a string naming the
+		/// type.
+		/// </summary>
+		/// <param name="type"></param>
+		/// <returns>An instance of <c>NHibernate.Type.IType</c></returns>
+		/// <remarks>
+		/// We check to see if it implements IType, ICompositeUserType, IUserType, ILifecycle (Association), or
+		/// IPersistentEnum.  If none of those are implemented then we will serialize the Type to the
+		/// database using NHibernate.Type.SerializableType(typeName)
+		/// </remarks>
+		public static IType HeuristicType(System.Type type)
+		{
+			// Use the basic name (such as String) to get the instance of the IType object.
+			if (typeByTypeOfName.TryGetValue(type.AssemblyQualifiedName, out var returnType))
```

Comment:
I see what you mean. But I think it's better to keep this logic once inside Basic method. I will try to refactor it